File: ee-schkola/src-gen/main/kotlin/ee/schkola/student/StudentApiBase.kt
package ee.schkola.student

import ee.schkola.SchkolaBase
import ee.schkola.Trace
import ee.schkola.person.Contact
import ee.schkola.person.PersonName
import ee.schkola.person.Profile
import java.util.*


enum class AttendanceState {
    REGISTERED,
    CONFIRMED,
    CANCELED,
    PRESENT;

    fun isRegistered() : Boolean = this == REGISTERED
    fun isConfirmed() : Boolean = this == CONFIRMED
    fun isCanceled() : Boolean = this == CANCELED
    fun isPresent() : Boolean = this == PRESENT
}


fun String?.toAttendanceState(): AttendanceState {
    return if (this != null) AttendanceState.valueOf(this) else AttendanceState.REGISTERED
}


enum class GroupCategory {
    COURSE_GROUP,
    YEAR_GROUP;

    fun isCourseGroup() : Boolean = this == COURSE_GROUP
    fun isYearGroup() : Boolean = this == YEAR_GROUP
}


fun String?.toGroupCategory(): GroupCategory {
    return if (this != null) GroupCategory.valueOf(this) else GroupCategory.COURSE_GROUP
}




open class Attendance : SchkolaBase {
    val student: Profile
    val date: Date
    val course: Course
    val hours: Int
    val state: AttendanceState
    val stateTrace: Trace
    val token: String
    val tokenTrace: Trace


    constructor(id: String = "", student: Profile = Profile(), date: Date = Date(), course: Course = Course(), hours: Int = 0, 
                state: AttendanceState = AttendanceState.REGISTERED, stateTrace: Trace = Trace(), token: String = "", 
                tokenTrace: Trace = Trace()) : super(id) {
        this.student = student
        this.date = date
        this.course = course
        this.hours = hours
        this.state = state
        this.stateTrace = stateTrace
        this.token = token
        this.tokenTrace = tokenTrace
    }

    companion object {
        val EMPTY = Attendance()
    }
}


open class Course : SchkolaBase {
    val name: String
    val begin: Date
    val end: Date
    val teacher: PersonName
    val schoolYear: SchoolYear
    val fee: Float
    val description: String


    constructor(id: String = "", name: String = "", begin: Date = Date(), end: Date = Date(), teacher: PersonName = PersonName(), 
                schoolYear: SchoolYear = SchoolYear(), fee: Float = 0f, description: String = "") : super(id) {
        this.name = name
        this.begin = begin
        this.end = end
        this.teacher = teacher
        this.schoolYear = schoolYear
        this.fee = fee
        this.description = description
    }

    companion object {
        val EMPTY = Course()
    }
}


open class Grade : SchkolaBase {
    val student: Profile
    val course: Course
    val grade: Float
    val gradeTrace: Trace
    val comment: String


    constructor(id: String = "", student: Profile = Profile(), course: Course = Course(), grade: Float = 0f, gradeTrace: Trace = Trace(), 
                comment: String = "") : super(id) {
        this.student = student
        this.course = course
        this.grade = grade
        this.gradeTrace = gradeTrace
        this.comment = comment
    }

    companion object {
        val EMPTY = Grade()
    }
}


open class Group : SchkolaBase {
    val name: String
    val category: GroupCategory
    val schoolYear: SchoolYear
    val representative: Profile
    val students: MutableList<Course>
    val courses: MutableList<Course>


    constructor(id: String = "", name: String = "", category: GroupCategory = GroupCategory.COURSE_GROUP, 
                schoolYear: SchoolYear = SchoolYear(), representative: Profile = Profile(), 
                students: MutableList<Course> = arrayListOf(), courses: MutableList<Course> = arrayListOf()) : super(id) {
        this.name = name
        this.category = category
        this.schoolYear = schoolYear
        this.representative = representative
        this.students = students
        this.courses = courses
    }

    companion object {
        val EMPTY = Group()
    }
}


open class SchoolApplication : SchkolaBase {
    val profile: Profile
    val recommendationOf: PersonName
    val churchContactPerson: PersonName
    val churchContact: Contact
    val schoolYear: SchoolYear
    val group: String


    constructor(id: String = "", profile: Profile = Profile(), recommendationOf: PersonName = PersonName(), 
                churchContactPerson: PersonName = PersonName(), churchContact: Contact = Contact(), 
                schoolYear: SchoolYear = SchoolYear(), group: String = "") : super(id) {
        this.profile = profile
        this.recommendationOf = recommendationOf
        this.churchContactPerson = churchContactPerson
        this.churchContact = churchContact
        this.schoolYear = schoolYear
        this.group = group
    }

    companion object {
        val EMPTY = SchoolApplication()
    }
}


open class SchoolYear : SchkolaBase {
    val name: String
    val start: Date
    val end: Date
    val dates: MutableList<Course>


    constructor(id: String = "", name: String = "", start: Date = Date(), end: Date = Date(), dates: MutableList<Course> = arrayListOf()) : super(id) {
        this.name = name
        this.start = start
        this.end = end
        this.dates = dates
    }

    companion object {
        val EMPTY = SchoolYear()
    }
}

File: ee-schkola/src-gen/main/go/ee/schkola/student/StudentApiBase.go
package student

import (
    "ee/schkola"
    "ee/schkola/person"
    "github.com/eugeis/gee/enum"
    "time"
)

type Attendance struct {
    Student  *person.Profile
    date *time.Time
    Course  *Course
    hours int
    state *AttendanceState
    stateTrace *schkola.Trace
    token string
    tokenTrace *schkola.Trace
    *schkola.SchkolaBase
}

func NewAttendance(student *person.Profile, date *time.Time, course *Course, hours int, state *AttendanceState, stateTrace *schkola.Trace, 
                token string, tokenTrace *schkola.Trace, SchkolaBase *schkola.SchkolaBase) (ret *Attendance) {
    ret = &Attendance{
        Student : student,
        date: date,
        Course : course,
        hours: hours,
        state: state,
        stateTrace: stateTrace,
        token: token,
        tokenTrace: tokenTrace,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type Course struct {
    name string
    begin *time.Time
    end *time.Time
    teacher *person.PersonName
    schoolYear *SchoolYear
    fee float64
    description string
    *schkola.SchkolaBase
}

func NewCourse(name string, begin *time.Time, end *time.Time, teacher *person.PersonName, schoolYear *SchoolYear, fee float64, 
                description string, SchkolaBase *schkola.SchkolaBase) (ret *Course) {
    ret = &Course{
        name: name,
        begin: begin,
        end: end,
        teacher: teacher,
        schoolYear: schoolYear,
        fee: fee,
        description: description,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type Grade struct {
    student *person.Profile
    course *Course
    grade float64
    gradeTrace *schkola.Trace
    comment string
    *schkola.SchkolaBase
}

func NewGrade(student *person.Profile, course *Course, grade float64, gradeTrace *schkola.Trace, comment string, 
                SchkolaBase *schkola.SchkolaBase) (ret *Grade) {
    ret = &Grade{
        student: student,
        course: course,
        grade: grade,
        gradeTrace: gradeTrace,
        comment: comment,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type Group struct {
    name string
    category *GroupCategory
    schoolYear *SchoolYear
    representative *person.Profile
    students []*Course
    courses []*Course
    *schkola.SchkolaBase
}

func NewGroup(name string, category *GroupCategory, schoolYear *SchoolYear, representative *person.Profile, students []*Course, 
                courses []*Course, SchkolaBase *schkola.SchkolaBase) (ret *Group) {
    ret = &Group{
        name: name,
        category: category,
        schoolYear: schoolYear,
        representative: representative,
        students: students,
        courses: courses,
        SchkolaBase: SchkolaBase,
    }
    
    return
}

func (o *Group) AddToStudents(item *Course) *Course {
    o.students = append(o.students, item)
    return item
}

func (o *Group) AddToCourses(item *Course) *Course {
    o.courses = append(o.courses, item)
    return item
}



type SchoolApplication struct {
    profile *person.Profile
    recommendationOf *person.PersonName
    churchContactPerson *person.PersonName
    churchContact *person.Contact
    schoolYear *SchoolYear
    group string
    *schkola.SchkolaBase
}

func NewSchoolApplication(profile *person.Profile, recommendationOf *person.PersonName, churchContactPerson *person.PersonName, 
                churchContact *person.Contact, schoolYear *SchoolYear, group string, SchkolaBase *schkola.SchkolaBase) (ret *SchoolApplication) {
    ret = &SchoolApplication{
        profile: profile,
        recommendationOf: recommendationOf,
        churchContactPerson: churchContactPerson,
        churchContact: churchContact,
        schoolYear: schoolYear,
        group: group,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type SchoolYear struct {
    name string
    start *time.Time
    end *time.Time
    dates []*Course
    *schkola.SchkolaBase
}

func NewSchoolYear(name string, start *time.Time, end *time.Time, dates []*Course, SchkolaBase *schkola.SchkolaBase) (ret *SchoolYear) {
    ret = &SchoolYear{
        name: name,
        start: start,
        end: end,
        dates: dates,
        SchkolaBase: SchkolaBase,
    }
    
    return
}

func (o *SchoolYear) AddToDates(item *Course) *Course {
    o.dates = append(o.dates, item)
    return item
}








type AttendanceState struct {
	name  string
	ordinal int
}

func (o *AttendanceState) Name() string {
    return o.name
}

func (o *AttendanceState) Ordinal() int {
    return o.ordinal
}

func (o *AttendanceState) IsRegistered() bool {
    return o == _attendanceStates.Registered()
}

func (o *AttendanceState) IsConfirmed() bool {
    return o == _attendanceStates.Confirmed()
}

func (o *AttendanceState) IsCanceled() bool {
    return o == _attendanceStates.Canceled()
}

func (o *AttendanceState) IsPresent() bool {
    return o == _attendanceStates.Present()
}

type attendanceStates struct {
	values []*AttendanceState
    literals []enum.Literal
}

var _attendanceStates = &attendanceStates{values: []*AttendanceState{
    {name: "Registered", ordinal: 0},
    {name: "Confirmed", ordinal: 1},
    {name: "Canceled", ordinal: 2},
    {name: "Present", ordinal: 3}},
}

func AttendanceStates() *attendanceStates {
	return _attendanceStates
}

func (o *attendanceStates) Values() []*AttendanceState {
	return o.values
}

func (o *attendanceStates) Literals() []enum.Literal {
	if o.literals == nil {
		o.literals = make([]enum.Literal, len(o.values))
		for i, item := range o.values {
			o.literals[i] = item
		}
	}
	return o.literals
}

func (o *attendanceStates) Registered() *AttendanceState {
    return _attendanceStates.values[0]
}

func (o *attendanceStates) Confirmed() *AttendanceState {
    return _attendanceStates.values[1]
}

func (o *attendanceStates) Canceled() *AttendanceState {
    return _attendanceStates.values[2]
}

func (o *attendanceStates) Present() *AttendanceState {
    return _attendanceStates.values[3]
}

func (o *attendanceStates) ParseAttendanceState(name string) (ret *AttendanceState, ok bool) {
	if item, ok := enum.Parse(name, o.literals); ok {
		return item.(*AttendanceState), ok
	}
	return
}


type GroupCategory struct {
	name  string
	ordinal int
}

func (o *GroupCategory) Name() string {
    return o.name
}

func (o *GroupCategory) Ordinal() int {
    return o.ordinal
}

func (o *GroupCategory) IsCourseGroup() bool {
    return o == _groupCategorys.CourseGroup()
}

func (o *GroupCategory) IsYearGroup() bool {
    return o == _groupCategorys.YearGroup()
}

type groupCategorys struct {
	values []*GroupCategory
    literals []enum.Literal
}

var _groupCategorys = &groupCategorys{values: []*GroupCategory{
    {name: "CourseGroup", ordinal: 0},
    {name: "YearGroup", ordinal: 1}},
}

func GroupCategorys() *groupCategorys {
	return _groupCategorys
}

func (o *groupCategorys) Values() []*GroupCategory {
	return o.values
}

func (o *groupCategorys) Literals() []enum.Literal {
	if o.literals == nil {
		o.literals = make([]enum.Literal, len(o.values))
		for i, item := range o.values {
			o.literals[i] = item
		}
	}
	return o.literals
}

func (o *groupCategorys) CourseGroup() *GroupCategory {
    return _groupCategorys.values[0]
}

func (o *groupCategorys) YearGroup() *GroupCategory {
    return _groupCategorys.values[1]
}

func (o *groupCategorys) ParseGroupCategory(name string) (ret *GroupCategory, ok bool) {
	if item, ok := enum.Parse(name, o.literals); ok {
		return item.(*GroupCategory), ok
	}
	return
}



